virtualizarr/manifests/store.py

```diff
@@ -0,0 +1,377 @@
+    """
+    # Start with expected group level metadata
+    raise NotImplementedError
+
+
+def get_zarr_metadata(manifest_group: ManifestGroup, key: str) -> Buffer:
+    """
+    Generate the expected Zarr V3 metadata from a virtual dataset.
+
+    Group metadata is returned for all Datasets and Array metadata
+    is returned for all DataArrays.
+
+    Combines the ManifestArray metadata with the attrs from the DataArray
+    and adds `dimension_names` for all arrays if not already provided.
+
+    Parameters
+    ----------
+    manifest_group : ManifestGroup
+    key : str
+
+    Returns
+    -------
+    Buffer
+    """
+    # If requesting the root metadata, return the standard group metadata with additional dataset specific attributes
+
+    if key == "zarr.json":
+        metadata = manifest_group._metadata.to_dict()
+        return dict_to_buffer(metadata, prototype=default_buffer_prototype())
+    else:
+        var, _ = key.split("/")
+        metadata = manifest_group._manifest_arrays[var].metadata.to_dict()
+        return dict_to_buffer(metadata, prototype=default_buffer_prototype())
+
+
+def parse_manifest_index(
+    key: str, chunk_key_encoding: str = "."
+) -> tuple[str, tuple[int, ...]]:
+    """
+    Splits `key` provided to a zarr store into the variable indicated
+    by the first part and the chunk index from the 3rd through last parts,
+    which can be used to index into the ndarrays containing paths, offsets,
+    and lengths in ManifestArrays.
+
+    Currently only works for 1d+ arrays with a tree depth of one from the
+    root Zarr group.
+
+    Parameters
+    ----------
+    key : str
+    chunk_key_encoding : str
+
+    Returns
+    -------
+    ManifestIndex
+    """
+    parts = key.split("/")
+    var = parts[0]
+    # Assume "c" is the second part
+    # TODO: Handle scalar array case with "c" holds the data
+    if chunk_key_encoding == "/":
+        indexes = tuple(int(ind) for ind in parts[2:])
+    else:
+        indexes = tuple(int(ind) for ind in parts[2].split(chunk_key_encoding))
+    return var, indexes
+
+
+def find_matching_store(stores: StoreDict, request_key: str) -> StoreRequest:
+    """
+    Find the matching store based on the store keys and the beginning of the URI strings,
+    to fetch data from the appropriately configured ObjectStore.
+
+    Parameters:
+    -----------
+    stores : StoreDict
+        A dictionary with URI prefixes for different stores as keys
+    request_key : str
+        A string to match against the dictionary keys
+
+    Returns:
+    --------
+    StoreRequest
+    """
+    # Sort keys by length in descending order to ensure longer, more specific matches take precedence
+    sorted_keys = sorted(stores.keys(), key=len, reverse=True)
+
+    # Check each key to see if it's a prefix of the uri_string
+    for key in sorted_keys:
+        if request_key.startswith(key):
+            return StoreRequest(store=stores[key], key=request_key[len(key) :])
```

Comment:
I think using `key=request_key[len(key) :]` makes some assumptions that are not universally valid. I will look into this tomorrow.